Mathlib/Analysis/PSeries.lean

```diff
@@ -19,134 +19,183 @@ converges if and only if so does `∑ k, 2 ^ k f (2 ^ k)`. We prove this test in
+    (hu : Monotone u) (n : ℕ) : (∑ k in range (u n), f k) ≤ ∑ k in range (u 0), f k +
+    ∑ k in range n, (u (k + 1) - u k) • f (u k) := by
+  convert add_le_add_left (le_sum_schlomilch' hf h_pos hu n) (∑ k in range (u 0), f k)
+  rw [← sum_range_add_sum_Ico _ (hu n.zero_le)]
 
-theorem sum_condensed_le' (hf : ∀ ⦃m n⦄, 1 < m → m ≤ n → f n ≤ f m) (n : ℕ) :
-    (∑ k in range n, 2 ^ k • f (2 ^ (k + 1))) ≤ ∑ k in Ico 2 (2 ^ n + 1), f k := by
+theorem sum_schlomilch_le' (hf : ∀ ⦃m n⦄, 1 < m → m ≤ n → f n ≤ f m) (h_pos : ∀ n, 0 < u n)
+    (hu : Monotone u) (n : ℕ) : (∑ k in range n, (u (k + 1) - u k) • f (u (k + 1))) ≤
+    ∑ k in Ico (u 0 + 1) (u n + 1), f k := by
   induction' n with n ihn
   · simp
-  suffices 2 ^ n • f (2 ^ (n + 1)) ≤ ∑ k in Ico (2 ^ n + 1) (2 ^ (n + 1) + 1), f k by
+  suffices (u (n + 1) - u n) • f (u (n + 1)) ≤ ∑ k in Ico (u n + 1) (u (n + 1) + 1), f k by
     rw [sum_range_succ, ← sum_Ico_consecutive]
     exacts [add_le_add ihn this,
-      (add_le_add_right n.one_le_two_pow _ : 1 + 1 ≤ 2 ^ n + 1),
-      add_le_add_right (Nat.pow_le_pow_of_le_right zero_lt_two n.le_succ) _]
-  have : ∀ k ∈ Ico (2 ^ n + 1) (2 ^ (n + 1) + 1), f (2 ^ (n + 1)) ≤ f k := by
-    -- Note(kmill): was `fun k hk => ...` but `mem_Ico.mp hk` was elaborating with some
-    -- delayed assignment metavariables that weren't resolved in time. `intro` fixes this.
-    intro k hk
-    exact hf (Nat.one_le_two_pow.trans_lt <| (Nat.lt_succ_of_le le_rfl).trans_le (mem_Ico.mp hk).1)
-      (Nat.le_of_lt_succ <| (mem_Ico.mp hk).2)
+      (add_le_add_right (hu n.zero_le) _ : u 0 + 1 ≤ u n + 1),
+      add_le_add_right (hu n.le_succ) _]
+  have : ∀ k ∈ Ico (u n + 1) (u (n + 1) + 1), f (u (n + 1)) ≤ f k := fun k hk =>
+    hf (Nat.lt_of_le_of_lt (Nat.succ_le_of_lt (h_pos n)) <| (Nat.lt_succ_of_le le_rfl).trans_le
+      (mem_Ico.mp hk).1) (Nat.le_of_lt_succ <| (mem_Ico.mp hk).2)
   convert sum_le_sum this
   simp [pow_succ, two_mul]
-#align finset.sum_condensed_le' Finset.sum_condensed_le'
-
-theorem sum_condensed_le (hf : ∀ ⦃m n⦄, 1 < m → m ≤ n → f n ≤ f m) (n : ℕ) :
-    (∑ k in range (n + 1), 2 ^ k • f (2 ^ k)) ≤ f 1 + 2 • ∑ k in Ico 2 (2 ^ n + 1), f k := by
-  convert add_le_add_left (nsmul_le_nsmul_right (sum_condensed_le' hf n) 2) (f 1)
-  simp [sum_range_succ', add_comm, pow_succ, mul_nsmul', sum_nsmul]
-#align finset.sum_condensed_le Finset.sum_condensed_le
 
+theorem sum_schlomilch_le {C : ℕ} (hf : ∀ ⦃m n⦄, 1 < m → m ≤ n → f n ≤ f m) (h_pos : ∀ n, 0 < u n)
+    (h_nonneg : ∀ n, 0 <= f n) (hu : Monotone u) (h_succ_diff : SuccDiffBounded C u) (n : ℕ) :
+    ∑ k in range (n + 1), (u (k + 1) - u k) • f (u k) ≤
+    (u 1 - u 0) • f (u 0) + C • ∑ k in Ico (u 0 + 1) (u n + 1), f k := by
+  rw [sum_range_succ', add_comm]
+  refine' add_le_add_left _ ((u 1 - u 0) • f (u 0))
+  suffices ∑ k in range n, (u (k + 2) - u (k + 1)) • f (u (k + 1)) ≤
+  C • ∑ k in range n, ((u (k + 1) - u k) • f (u (k + 1))) by
+    refine' le_trans this _
+    apply nsmul_le_nsmul_right
```

Comment:
```suggestion
    refine this.trans (nsmul_le_nsmul_right ?_ _)
```
Generally, `refine'` is discouraged, the better alternatives are `apply` (if you don't want to use tons of trailing underscores) and `refine` (where you have to specify with `?_` which goals are still left)

Reply:
I believe it's not because of performance: basically, they are very similar; `refine` is clearer (as it distinguishes so-called holes (`_` in the code) which you want Lean to fill from `?_` which you want Lean to turn into new goals for you.
This [thread](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/refine.20vs.20refine') on zulip goes into many details. In particular, there are about eight thousand calls of `refine'` left, so these will not be all rewritten manually.